vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/dataview/CheckboxGroupListDataViewImpl.java

```diff
@@ -0,0 +1,152 @@
+        return getNextOrPreviousItem(item, itemsList -> {
+            int itemIndex = itemsList.indexOf(item);
+            if (itemIndex != -1 && itemIndex < itemsList.size() - 1) {
+                return itemsList.get(itemIndex + 1);
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public boolean hasNextItem(T item) {
+        return getNextItem(item) != null;
+    }
+
+    @Override
+    public T getPreviousItem(T item) {
+        return getNextOrPreviousItem(item, itemsList -> {
+            int itemIndex = itemsList.indexOf(item);
+            if (itemIndex > 0) {
+                return itemsList.get(itemIndex - 1);
+            }
+            return null;
+        });
+    }
+
+    @Override
+    public boolean hasPreviousItem(T item) {
+        return getPreviousItem(item) != null;
+    }
+
+    @Override
+    public CheckboxGroupListDataView<T> withFilter(SerializablePredicate<T> filter) {
+        CheckboxGroupListDataView<T> dataView = withFilterOrOrder(
+                dataProvider -> dataProvider.setFilter(filter));
+        // fire size change event if size has changed
+        getDataSize();
+        return dataView;
+    }
+
+    @Override
+    public CheckboxGroupListDataView<T> withSortComparator(SerializableComparator<T> sortComparator) {
+        return withFilterOrOrder(dataProvider -> dataProvider.setSortComparator(sortComparator));
+    }
+
+    @Override
+    public Stream<T> getAllItems() {
+        DataProvider<T, ?> dataProvider = getDataProvider();
+        return dataProvider.fetch(new Query<>());
+    }
+
+    @Override
+    public int getDataSize() {
+        int size = getDataProvider().size(new Query<>());
+        if (size != itemsSize && sizeChangeListeners != null) {
+            sizeChangeListeners.forEach(listener -> listener.sizeChanged(
+                    new SizeChangeEvent<>(this.component, size)));
+            itemsSize = size;
+        }
+        return size;
+    }
+
+    @Override
+    public boolean dataContainsItem(T item) {
+        // InMemoryDataProviders do not override getId(), so rely on equals() method
+        return getAllItems().collect(Collectors.toList()).contains(item);
+    }
+
+    @Override
+    public T getItemOnRow(int row) {
+        if (row < 0) {
+            throw new IndexOutOfBoundsException("Row number should be zero on greater");
+        }
+
+        List<T> itemsList = getAllItems().collect(Collectors.toList());
+
+        if (itemsList.isEmpty()) {
+            throw new IndexOutOfBoundsException("Item requested on an empty data set");
+        }
+
+        final int itemsSize = itemsList.size();
+        if (row >= itemsSize) {
+            throw new IndexOutOfBoundsException(
+                    String.format("Row number is outside of data set size, row=%d, data set size=%d", row, itemsSize));
+        }
+
+        return itemsList.get(row);
+    }
+
+    // This could be a common method on top of List- and Lazy- DataViews
```

Comment:
This should probably be in an AbstractDataView in Flow data.

Reply:
Done.